File: Lua/Framework/LinkedList/LinkedList.lua
local LinkedListNode = require("LinkedListNode")

---@class LinkedList
---@field new fun():LinkedList
local LinkedList = sealedClass("LinkedList")

function LinkedList:ctor()
    self.Count = 0;
    self.m_Null = LinkedListNode.new()
    self.m_Null.m_Next = self.m_Null
    self.m_Null.m_Prev = self.m_Null
end

---@return LinkedListNode
function LinkedList:GetFirst()
    return self.m_Null.m_Next
end

---@return LinkedListNode
function LinkedList:GetLast()
    return self.m_Null.m_Prev
end

---@return LinkedListNode
function LinkedList:GetEnd()
    return self.m_Null
end

---@param node LinkedListNode
function LinkedList:PushBack(node)
    local prev = self.m_Null.m_Prev

    prev.m_Next = node
    node.m_Prev = prev

    node.m_Next = self.m_Null
    self.m_Null.m_Prev = node

    self.Count = self.Count + 1;
end

---@return LinkedListNode
function LinkedList:PopBack()
    if not self:IsEmpty() then
        local node = self.m_Null.m_Prev
        local prev = node.m_Prev
    
        prev.m_Next = self.m_Null
        self.m_Null.m_Prev = prev
    
        node.m_Prev = nil
        node.m_Next = nil
    
        self.Count = self.Count - 1;
        return node
    end
    
    return nil
end

---@param node LinkedListNode
function LinkedList:PushFront(node)
    local next = self.m_Null.m_Next
    
    next.m_Prev = node
    node.m_Next = next
    
    node.m_Prev = self.m_Null
    self.m_Null.m_Next = node
    self.Count = self.Count + 1;
end

---@return LinkedListNode
function LinkedList:PopFront()
    if not self:IsEmpty() then
        local node = self.m_Null.m_Next
        local next = node.m_Next

        next.m_Prev = self.m_Null
        self.m_Null.m_Next = next

        node.m_Prev = nil
        node.m_Next = nil

        self.Count = self.Count - 1;
        return node
    end
    
    return nil
end

---@param node LinkedListNode
function LinkedList: Remove(node)
    local prev = node.m_Prev
    local next = node.m_Next
    
    prev.m_Next = next
    next.m_Prev = prev
    
    node.m_Prev = nil
    node.m_Next = nil
    self.Count = self.Count - 1;
end

---@return boolean 
function LinkedList:IsEmpty()
    return (self.m_Null.m_Prev == self.m_Null and self.m_Null.m_Next == self.m_Null)
end

function LinkedList:Clear()
    self.m_Null.m_Next = self.m_Null
    self.m_Null.m_Prev = self.m_Null
end

return LinkedList

File: Packages/com.funplus.dragonreborn@1.0.0/Runtime/Structure/LinkedList.cs
 namespace DragonReborn
 {
     public class LinkedList
     {
         public class Node
         {
             public Node Prev;
             public Node Next;
         }

         private readonly Node _null = new Node();

         public LinkedList()
         {
             _null.Next = _null;
             _null.Prev = _null;
         }

         public Node GetFirst()
         {
             return _null.Next;
         }

         public Node GetLast()
         {
             return _null.Prev;
         }

         public Node GetEnd()
         {
             return _null;
         }

         public void PushBack(Node node)
         {
             var prev = _null.Prev;
		
             prev.Next = node;
             node.Prev = prev;
		
             node.Next = _null;
             _null.Prev = node;
         }

         public Node PopBack()
         {
             if (IsEmpty())
             {
                 return null;
             }
             
             var node = _null.Prev;
             var prev = node.Prev;

             prev.Next = _null;
             _null.Prev = prev;

             node.Prev = null;
             node.Next = null;

             return node;
         }
	
         public void PushFront(Node node)
         {
             var next = _null.Next;
		
             next.Prev = node;
             node.Next = next;
		
             node.Prev = _null;
             _null.Next = node;
         }

         public Node PopFront()
         {
             if (IsEmpty())
             {
                 return null;
             }
             
             var node = _null.Next;
             var next = node.Next;

             next.Prev = _null;
             _null.Next = next;

             node.Prev = null;
             node.Next = null;

             return node;
         }
	
         public void Remove(Node node)
         {
             var prev = node.Prev;
             var next = node.Next;
		
             prev.Next = next;
             next.Prev = prev;
		
             node.Prev = null;
             node.Next = null;
         }

         public bool IsEmpty()
         {
             return _null.Prev == _null && _null.Next == _null;
         }
     }
 }